### PR TITLE
fix: [SVS-02C] Inefficient mapping Lookups

### LINKED ISSUE #89: [SVS-02C] Inefficient `mapping` Lookups
## SVS-02C: Inefficient `mapping` Lookups

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Gas Optimization](/appendix/finding-types#gas-optimization) |  | [SortedVessels.sol:L143](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L143), [L147-L148](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L147-L148), [L151-L153](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L151-L153), [L156-L158](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L156-L158), [L161-L164](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L161-L164), [L167](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L167), [L184](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L184), [L186](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L186), [L189](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L189), [L191](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L191), [L195](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L195), [L197](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L197), [L201-L203](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L201-L203), [L205-L207](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L205-L207), [L212-L213](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L212-L213), [L216-L217](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L216-L217), [L370](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L370), [L376](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L376), [L383-L384](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L383-L384), [L404](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L404), [L410](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L410), [L417-L418](https://github.com/Gravita-Protocol/Gravita-SmartContracts/blob/bfa97cb37dc0e2927e75b36753585140f25a26dd/contracts/SortedVessels.sol#L417-L418) |

### Description:

The linked statements perform key-based lookup operations on `mapping` declarations from storage multiple times for the same key redundantly.

### Example:

```solidity
data[_asset].nodes[_id].exists = true;

if (prevId == address(0) && nextId == address(0)) {
	// Insert as head and tail
	data[_asset].head = _id;
	data[_asset].tail = _id;
} else if (prevId == address(0)) {
	// Insert before `prevId` as the head
	data[_asset].nodes[_id].nextId = data[_asset].head;
	data[_asset].nodes[data[_asset].head].prevId = _id;
	data[_asset].head = _id;
} else if (nextId == address(0)) {
	// Insert after `nextId` as the tail
	data[_asset].nodes[_id].prevId = data[_asset].tail;
	data[_asset].nodes[data[_asset].tail].nextId = _id;
	data[_asset].tail = _id;
} else {
	// Insert at insert position between `prevId` and `nextId`
	data[_asset].nodes[_id].nextId = nextId;
	data[_asset].nodes[_id].prevId = prevId;
	data[_asset].nodes[prevId].nextId = _id;
	data[_asset].nodes[nextId].prevId = _id;
}

data[_asset].size = data[_asset].size.add(1);
```

### Recommendation:

As the lookups internally perform an expensive `keccak256` operation, we advise the lookups to be cached wherever possible to a single local declaration that either holds the value of the `mapping` in case of primitive types or holds a `storage` pointer to the `struct` contained.

### PR DESCRIPTION
**Pull request by gbirckan**
fixes #89

### CODE REVIEW

File: contracts/SortedVessels.sol [hunk 1]
```diff
@@ -360,23 +354,20 @@ contract SortedVessels is OwnableUpgradeable, ISortedVessels {
...
 
 		// Descend the list until we reach the end or until we find a valid insert position
-		while (
-			prevId != address(0) &&
-			!_validInsertPosition(_asset, _vesselManager, _NICR, prevId, nextId)
-		) {
-			prevId = data[_asset].nodes[prevId].nextId;
-			nextId = data[_asset].nodes[prevId].nextId;
+		while (prevId != address(0) && !_validInsertPosition(_asset, _vesselManager, _NICR, prevId, nextId)) {
+			prevId = assetData.nodes[prevId].nextId;
```
Comment (0xfornax) [inline contracts/SortedVessels.sol:369]: Can be improved (assetData.nodes[prevId] repeated)

File: contracts/SortedVessels.sol [hunk 2]
```diff
@@ -360,23 +354,20 @@ contract SortedVessels is OwnableUpgradeable, ISortedVessels {
...
+			nextId = assetData.nodes[prevId].nextId;
```
Comment (0xfornax) [inline contracts/SortedVessels.sol:370]: Can be improved (assetData.nodes[prevId] repeated)

File: contracts/SortedVessels.sol [hunk 3]
```diff
@@ -394,23 +385,20 @@ contract SortedVessels is OwnableUpgradeable, ISortedVessels {
...
 		// Ascend the list until we reach the end or until we find a valid insertion point
-		while (
-			nextId != address(0) &&
-			!_validInsertPosition(_asset, _vesselManager, _NICR, prevId, nextId)
-		) {
-			nextId = data[_asset].nodes[nextId].prevId;
-			prevId = data[_asset].nodes[nextId].prevId;
+		while (nextId != address(0) && !_validInsertPosition(_asset, _vesselManager, _NICR, prevId, nextId)) {
+			nextId = assetData.nodes[nextId].prevId;
+			prevId = assetData.nodes[nextId].prevId;
```
Comment (0xfornax) [inline contracts/SortedVessels.sol:401]: Same here.

File: contracts/SortedVessels.sol (hunk 2 above)
Comment (gbirckan) [reply contracts/SortedVessels.sol:370]: Nope, look closely and you'll notice that prevId changes between those two lines, so we cannot cache the lookup.

File: contracts/SortedVessels.sol (hunk 3 above)
Comment (gbirckan) [reply contracts/SortedVessels.sol:401]: Nope, look closely and you'll notice that prevId changes between those two lines, so we cannot cache the lookup.

**Review by 0xfornax — APPROVED** [review]
lgtm

**Pull request merged**